python/tvm/topi/hexagon/qnn/avg_pool2d.py

```diff
@@ -0,0 +1,205 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, unused-variable, unused-argument, too-many-locals
+
+""" Compute and schedule for quantized avg_pool2d op
+
+Please note the following assumptions made by the implementation:
+
+1) The input must be padded in advance to account for 'padding'. In addition,
+   both input and output must be padded as per the physical buffer layout.
+2) The current implementation assumes 'count_include_pad' to be 'True'. It can be
+   modified to support 'False' case but the element count for the pooling window
+   must be pre-computed and provided as an input to reduce the run-time overhead.
+3) 'padding' is ignored. It must be handled outside of the sliced op.
+4) Please note that this implementation will not work if the output includes any
+   physical layout related padding as it can result into out-of-bound access
+   for the input.
+"""
+
+from tvm import te
+from tvm import tir
+from ..utils import get_layout_transform_fn, get_fixed_point_value
+
+
+def validate_out_shape(out_shape: list, in_shape: list, kernel: list, stride: list, dilation: list):
+    """Validate output shape"""
+    _, oh, ow, _ = out_shape
+    _, ih, iw, _ = in_shape
+    kh, kw = kernel
+    sh, sw = stride
+    dh, dw = dilation
+    if ih < (oh - 1) * sh + dh * (kh - 1) + 1:
+        raise RuntimeError("Output height is too large")
+    if iw < (ow - 1) * sw + dw * (kw - 1) + 1:
+        raise RuntimeError("Output width is too large")
+
+
+def saturate(x: te.Tensor, dtype: str):
+    """Saturate value for the specified data type"""
+    return te.max(te.min_value(dtype), te.min(x, te.max_value(dtype)))
```

Comment:
When I looked at several of the Hexagon `.so` files produced by this PR's unit tests, I didn't see any indication that Hexagon's `saturate` or `:sat` instructions were being used.

This isn't a critique of the PR; I'm just mentioning it as a point of interest for future work.

Reply:
Actually, I'm wondering if the `saturate` can sometimes be elided entirely when dtype=float16.

Here's my (maybe flawed) reasoning:
- Hexagon has two units that might do this math: the ARM core, and the HVX units.
- If this code runs on the ARM core, then we it's treated as an IEEE-754 single-precision float.
- If this code runs on an HVX core, then it's going to be processed using qfloat16 semantics, which automatically uses saturate behavior.

So any dataflow path that _definitely_ involves qfloat16 representation could (perhaps) entirely avoid explicit saturation logic.

I'm starting to wonder if TIR should eventually distinguish saturated vs. unsaturated ops.

Reply:
Thanks for the comment, @cconvey! You're correct about saturate not being needed for float16 dtype.  Please note that the functions in this file qnn/avg_pool2d.py are meant to be used only for the quantized models and therefore should have uint8 and int8 dtypes.

Reply:
> When I looked at several of the Hexagon `.so` files produced by this PR's unit tests, I didn't see any indication that Hexagon's `saturate` or `:sat` instructions were being used.
> 
> This isn't a critique of the PR; I'm just mentioning it as a point of interest for future work.

That's very likely. Thanks for looking into it!

Unless we generate saturating llvm instructions through TVM, we will have to add additional code in LLVM to recognize the sequence of min, max as saturate.

Reply:
I agree. I think it will be better to do it in TVM as we can generate the appropriate LLVM saturating instruction during TVM codegen which can then be lowered into target specific instructions in the LLVM backend.